feature(js): Adds a fixed Ajax spinner module

mrclay opened the pull request:
I think I based this on a Google site. A fixed spinner seems like a better general solution than figuring out where to put them within a page.
- [x] Do not use inline js
- [x] Document how to use the spinner
- [x] Test

mrclay committed:
feature(js): Adds a fixed Ajax spinner module

This also moves inline theme sandbox scripts to an AMD module

juho-jaakkola commented on mod/aalborg_theme/views/default/css/elements/misc.php:
```diff
@@ -62,3 +62,22 @@
+
+.elgg-wait-cursor,
+.elgg-wait-cursor * {
+	cursor:wait !important
```
Why does this need to use ```!important```? Usually the need to use it means that there's something wrong with the CSS elsewhere.

mrclay replied:
This was mostly pulled as is from a site. The important probably isn't necessary.

juho-jaakkola replied:
Might be best to check the other properties also and remove anything that's not needed by Elgg.

mrclay replied:
Will do. Style doesn't match either.

juho-jaakkola commented on views/default/css/elements/misc.php:
```diff
@@ -59,3 +59,5 @@
 	padding: 0;
 }
+
+<?php echo elgg_view('css/elements/spinner.css') ?>
```
Why is this not called in ```views/default/css/elgg```?

mrclay replied:
@ewinslow had brought up the idea of separating CSS into more files. For BC those would be pulled in from the original css/elgg/elements/*.php files, so this was for consistency with that plan I guess

juho-jaakkola replied:
Ok, I see. This file wouldn't introduce anything non-BC, though?

mrclay replied:
I'm not sure what you're asking.

juho-jaakkola replied:
I mean we could include this also in ```views/default/css/elgg```.

juho-jaakkola commented on mod/aalborg_theme/views/default/css/elements/misc.php:
```diff
@@ -62,3 +62,22 @@
+	border-radius:0 0 10px 10px;
+	border:1px solid #CCC;
+	border-top:0
+}
```
Why is this added to Aalborg directly, but in core through ```css/elements/misc```?

mrclay replied:
Mistake :)

mrclay commented:
Before I finish this are there any objections to adding something like this? On the site I pulled it from we even use it for some non-ajaxed form submissions as it makes the UX seem a bit more responsive before the full page redraw begins.

mrclay commented on views/default/js/elgg/spinner.js:
```diff
@@ -0,0 +1,66 @@
+		 * @param {jQuery} $wait       $-wrapped element(s) from which to remove the elgg-wait-cursor class
+		 * @param {Object} fadeOptions Options for the $.fadeOut method
+		 */
+		remove: function ($wait, fadeOptions) {
```
I could get rid of remove() if I returned a Promise ($.Deferred) that the user would manually resolve().

ewinslow commented on engine/lib/views.php:
```diff
@@ -1604,6 +1604,8 @@ function elgg_views_boot() {
 	$elgg_js_url = elgg_get_simplecache_url('js', 'elgg');
 	elgg_register_js('elgg', $elgg_js_url, 'head');
 
+	elgg_extend_view('js/elgg', 'js/elgg/spinner.js', 501);
```
Prefer not to extend this view any more if possible. I know it creates more http requests but this is not something that should block page rendering anyways so I think that's OK.

ewinslow commented on views/default/js/elgg/spinner.js:
```diff
@@ -0,0 +1,66 @@
+		num_running = 0,
+		$spinner;
+
+	var exports = {
```
Add,remove,remove all? Smells like a collection api! Not sure what to do about that right now, just sort of interesting to me.

ewinslow commented on views/default/js/elgg/spinner.js:
```diff
@@ -0,0 +1,66 @@
+define('elgg/spinner', function (require) {
```
Make this anonymous.

ewinslow commented on views/default/js/elgg/spinner.js:
```diff
@@ -0,0 +1,66 @@
+		 */
+		removeAll: function (fadeOptions) {
+			if (num_running) {
+				num_running = 1;
```
Is this supposed to be 0?

ewinslow commented on views/default/js/elgg/spinner.js:
```diff
@@ -0,0 +1,66 @@
+				if ($spinner) {
+					$spinner.show();
+				} else {
+					$spinner = $('<div class="elgg-spinner elgg-ajax-loader"></div>');
```
So is spinner a type of ajax loader?

ewinslow commented on views/default/js/elgg/spinner.js:
```diff
@@ -0,0 +1,66 @@
+				return;
+			}
+
+			if ($wait instanceof $) {
```
Can we just assert this as a precondition and throw if not met? I prefer not to add branches to our  code if possible.

ewinslow commented:
it would also be awesome if you could write a test for this.

Right now we're basically exporting a global object (global state). Have you considered exporting a class instead that gets instantiated elsewhere?

mrclay commented:
Converted to constructor and documented

mrclay commented on js/lib/ajax.js:
```diff
@@ -14,7 +14,7 @@ elgg.provide('elgg.ajax');
  * @param {string} url Optionally specify the url as the first argument
  * @param {Object} options Optional. {@link jQuery#ajax}
- * @return {XmlHttpRequest}
+ * @return {jqXHR}
```
You think this needs a separate PR?

ewinslow replied:
Separate commit maybe?

mrclay commented:
Added some tests!

ewinslow commented on js/tests/ElggSpinnerTest.js:
```diff
@@ -0,0 +1,68 @@
+
+		it("Indicator stays until all instances stopped", function() {
+			var s1 = new Spinner(),
+				s2 = new Spinner({
```
Tabs are making this not align well. Spaces for alignment?

ewinslow replied:
Alternatively, just declare a new var on each line and let the minifier compress them (my preference).

ewinslow commented on docs/guides/javascript.rst:
```diff
@@ -339,6 +339,30 @@ Calls an Elgg action with the data passed. This handles outputting of system mes
       }
    });
 
+Module elgg/Spinner
```
Code ticks?

ewinslow commented on docs/guides/javascript.rst:
```diff
@@ -339,6 +339,30 @@ Calls an Elgg action with the data passed. This handles outputting of system mes
+   });
+
+**Note:** The indicator is only removed once all Spinner instances are stopped.
+
```
Thank you for the docs!

hypeJunction replied:
In my code, I prefer using beforeSend parameter to display spinners. It seems more organized this way.

mrclay replied:
I changed it to use beforeSend, and switched this to reference the instance methods directly (also added a test to verify we can call these without `this` context)

mrclay commented:
All should be taken care on this.

ewinslow commented on docs/guides/javascript.rst:
```diff
@@ -339,6 +339,30 @@ Calls an Elgg action with the data passed. This handles outputting of system mes
    });
 
+Module ``elgg/Spinner``
+~~~~~~~~~~~~~~~~~~~~~~~
```
Seems like sphinx doesn't like this header notation. It may work to switch to all dashes

ewinslow commented on mod/developers/views/default/js/elgg/dev.js:
```diff
@@ -2,10 +2,11 @@
-		
+	require('jquery.jstree');
+
+	function setupInspectTree() {
```
This file can be or extracted to a separate commit/PR. It's cleanup that isn't relevant to this change.

ewinslow commented on views/default/css/admin.php:
```diff
@@ -1595,6 +1595,8 @@
 	margin: 0 0 1em 2em;
 }
 
+<?php echo elgg_view('css/elements/spinner.css') ?>
```
I don't think this CSS has the right view name.

ewinslow commented:
So I can't help but think that all the files for this component should be much more localized. It is a pretty standalone component but you literally had to edit 10+ files in very distant places to get it integrated.

Here's my overview of the pieces of this PR with the files that were modified.

* implementation
  * views/default/js/elgg/spinner.js (logic)
  * views/default/css/elements/misc/spinner.css (style)
* unit tests (js/tests/ElggSpinnerTest.js)
* developer docs (docs/guides/javascript.rst)
* demo
  * mod/developers/start.php
  * mod/developers/views/default/js/elgg/dev/theme_sandbox.js
  * mod/developers/views/default/theme_sandbox/javascript.php
  * mod/developers/views/default/theme_sandbox/javascript/spinner.php
* integration
  * views/default/css/admin.php
  * views/default/css/elgg.php

It feels like there is hardly any rhyme or reason to which files need editing. Shouldn't things be as simple as:

* Spinner.css -- styles
* Spinner.js -- implementation
* Spinner.rst -- dev docs
* SpinnerDemo.html -- demo templates
* SpinnerDemo.js -- demo scripts
* SpinnerTest.js -- unit tests

pick a folder, plop these all in, done.
??

mrclay commented on views/default/css/admin.php:
```diff
@@ -1595,6 +1595,8 @@
 	margin: 0 0 1em 2em;
 }
 
+<?php echo elgg_view('css/elements/misc/spinner.css') ?>
```
@ewinslow fixed link

mrclay commented:
@ewinslow Yes, JS components having a single location would be great. We might be able to lean on grunt for this, but it would add a build step for regular devs.

mrclay commented:
Since fadeOut is always async I found the JS tests weren't really working and were not run in isolation as I'd expected, so I rewrote them using [Jasmine's blocking functions](http://jasmine.github.io/1.3/introduction.html#section-Asynchronous_Support) and added a private module function to reset global state.

mrclay commented:
I no longer think this API is right for core. I'd prefer just start() and stop() methods on a singleton module. No $wait or fade options. Fade using CSS animation.

ewinslow commented:
Did you want to update and submit, or just close this PR?

mrclay commented:
I'll make another PR.

mrclay commented:
Replaced by #7741
